File: Assets/Scripts/WebSocket/ImageAssembler.py
import base64
import tempfile
from io import BytesIO
from PIL import Image


class ImageAssembler:
    def __init__(self):
        self.image_chunks = []
        self.total_chunks = 0
        self.received_chunks = 0

    def process_image_data(self, image_data):
        # Parse metadata
        if image_data.startswith("ImageChunks"):
            self.total_chunks = int(image_data.split(' ')[1])
            # print(f"Total chunks: {self.total_chunks}")
            return
        # Receive image data chunks
        if image_data.startswith("Base64EncodedChunk"):
            # Extract Base64 encoded image data
            base64_data = image_data.split(' ')[1]
            # print(f"Received chunk {self.received_chunks + 1}/{self.total_chunks}")
            # print(base64_data)
            # Ensure length is valid for decoding
            if len(base64_data) % 4 != 0:
                print(len(base64_data))
                # Add padding if necessary to make length a multiple of 4
                base64_data += '=' * (4 - len(base64_data) % 4)
            try:
                # Decode Base64 encoded image data to bytes
                chunk_data = base64.b64decode(base64_data)
                self.image_chunks.append(chunk_data)
                self.received_chunks += 1
            except Exception as e:
                print(f"Error decoding Base64 data: {e}")

        # If all chunks received, assemble image
        if self.received_chunks == self.total_chunks:
            self.assemble_image()
            self.received_chunks = 0
            self.total_chunks = 0

    def assemble_image(self):
        # Concatenate chunks
        image_data = b''.join(self.image_chunks)
        self.image_chunks.clear()
        try:
            # Reconstruct image from byte array
            reconstructed_image = Image.open(BytesIO(image_data))
            print("Image reconstructed")
            # Use tempfile to create a temporary file
            with tempfile.NamedTemporaryFile(suffix=".jpg", delete=False) as temp_file:
                temp_file_path = temp_file.name
                reconstructed_image.save(temp_file_path)

            # This is the image that needs to be used by the dialogue agent to determine the current step
            # Display the temporary image file - this is just for testing purposes
            reconstructed_image.show()
        except Exception as e:
            print(f"Error reconstructing image: {e}")

File: Assets/Scripts/WebSocket/ImageAssembler.cs
using UnityEngine;
using System;
using System.Collections.Generic;
using System.Drawing;
using System.IO;
using System.Text;
using System.Threading;
using UnityEngine.UI;
using System.Collections;

public class ImageAssembler : MonoBehaviour
{
    [SerializeField] RawImage rawImage; // Material to hold the reconstructed image

    [SerializeField]
    UdpSocket WebSocket;

    private List<byte[]> imageChunks = new List<byte[]>();
    private int totalChunks = 0;
    private int receivedChunks = 0;

    public float duration = 5f; // Timer duration in seconds
    private float timeRemaining;
    string resend = "";

    void Start()
    {
        timeRemaining = duration;
        StartCoroutine(TimerCoroutine());
    }

    IEnumerator TimerCoroutine()
    {
        while (timeRemaining > 0)
        {
            timeRemaining -= Time.deltaTime;
            yield return null; // Wait for the next frame
        }
        TimerEnded();
    }

    void TimerEnded()
    {
        resend = "Resend Image";
    }

    public string ProcessImageData(string imageData)
    {
        // Parse metadata
        if (imageData.StartsWith("ImageChunks"))
        {
            receivedChunks = 0;
            totalChunks = int.Parse(imageData.Split(' ')[1]);
            return "";
        }
        // Receive image data chunks
        if (imageData.StartsWith("Base64EncodedChunk"))
        {
            // Extract Base64 encoded image data
            string base64Data = imageData.Split(' ')[1];

            // Ensure length is valid for decoding
            if (base64Data.Length % 4 != 0)
            {
                // Add padding if necessary to make length a multiple of 4
                base64Data += new string('=', (4 - base64Data.Length % 4) % 4);
            }

            // Decode Base64 encoded image data to bytes
            byte[] chunkData = Convert.FromBase64String(base64Data);
            imageChunks.Add(chunkData);
            receivedChunks++;
        }
        // If all chunks received, assemble image
        if (totalChunks == 0)
        {
            return "Resend Image - Image Chunk";
        }
        else if (receivedChunks == totalChunks)
        {
            receivedChunks = 0;
            totalChunks = 0;
            Thread thread = new Thread(AssembleImage);
            thread.Start();
        }
        else if (resend != "") {
            resend = "";
            return "Resend Image - Timeout";
        }
        return "";

    }

    void AssembleImage()
    {

        UnityMainThreadDispatcher.Instance().Enqueue(() =>
        {
            byte[] imageData = CombineChunks(imageChunks);
            imageChunks.Clear();
            // Decode the byte image data into a Texture2D
            Texture2D texture = new Texture2D(1920, 1080);
            texture.LoadImage(imageData); // Load the byte image data

            //find ui game object
            if (rawImage == null)
            {
                rawImage = GameObject.Find("RawImage").GetComponent<RawImage>();
            }

            //set image transparency
            if (rawImage.texture == null)
            {
                var tempColor = rawImage.color;
                tempColor.a = 255f;
                rawImage.color = tempColor;
            }

            //Debug.Log($"Image loaded from {imageData.Length} bytes of data");

            rawImage.texture = texture;

            //Debug.Log("Image reconstructed and displayed on Quad.");
        });
    }

    private byte[] CombineChunks(List<byte[]> chunks)
    {
        using (MemoryStream stream = new MemoryStream())
        {
            foreach (byte[] chunk in chunks)
            {
                stream.Write(chunk, 0, chunk.Length);
            }
            return stream.ToArray();
        }
    }
}
